File: Lesson01/AlienFinder.cs
using Neo.Smartcontract.Framework;
using Neo.SmartContract.Framework.Services.Neo;
using System;
using System.Numerics;

public class AlienFinder : Smartcontract
{
    struct Alien
    {
        private uint xna;
        private string alienName;
        private uint blockHeight;

        public Alien(uint xna, string alienName, uint blockHeight) 
        {
            this.xna = xna; 
            this.alienName = alienName;
            this.blockHeight = blockHeight;
        }
    }

    // list to keep track of all Alien structs; 
    static List<Alien> aliens = new List<Alien>(); 

    public static void Main(string alienName) 
    {
        uint blockHeight = Blockchain.GetHeight();
        uint xna = FinaXnaOnPlanet(GeneratePlanet(blockHeight));
        Alien someAlien = new Alien(xna, alienName, blockHeight);
        aliens.Add(someAlien);
        Runtime.Notify("Alien created, ID: " + ToString(aliens.count - 1));
    }

    private static ulong GeneratePlanet(uint blockHeight)
    {
        return Blockchain.GetHeader(blockHeight).ConsensusData; 
    }

    private static uint FindXnaOnPlanet(ulong planet)
    {
        return (uint)(planet % 100000000);
    }
    
}

File: Typescript/en-us/AlienFinder.ts
import { Blockchain, constant, ArrayStorage, createEventNotifier, SmartContract } from '@neo-one/smart-contract';

const notifyCreation = createEventNotifier<number>(
    'Alien created, ID: ',
    'id'
);

export class AlienFinder extends SmartContract {

    // array to keep track of all Alien tuples; 
    private readonly aliens = ArrayStorage.for<[number, string, number]>();

    public createAlien(alienName: string) {
        let blockHeight: number = Blockchain.currentHeight;
        let xna: number = this.findXna(this.randomNumber(blockHeight));
        let someAlien: [number, string, number] = [xna, alienName, blockHeight];
        this.aliens.push(someAlien);
        notifyCreation(this.aliens.length - 1);
    }

    @constant
    private randomNumber(blockHeight: number): number {
        return Blockchain.currentBlockTime * blockHeight;
    }

    @constant
    private findXna(randomNumber: number): number {
        return randomNumber % 100000000;
    }
}